### DISCUSSION TITLE
Портирование на Эльбрус e2k

### DISCUSSION THREAD

**Discussion by EntityFX** [pascalabcnet/pascalabcnet#2411] [Q&A]
Возможно ли портировать сам компилятор PASCAL.NET на Эльбрус.

Рантайм Mono на Эльбрусе такой:

Mono JIT compiler version 5.16.0.220 (5.16.0.220/e2k_3.2 Thu Oct  8 05:49:17 GMT 2020)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       normal
	Notifications: epoll
	Architecture:  e2k
	Disabled:      none
	Misc:          
	Interpreter:   yes
	LLVM:          yes(3.6.0svn-mono-)
	GC:            sgen (concurrent by default)

Компилятор -- XBUILD.

Но можно собрать хоть на винде (и если там ANY CPU и нет зависимостей от нативных библиотек), то стартанёт на Эльбрус.

Я так понимаю, чисто компилятор заработать теоретически сможет. Иде -- нет. Можно использовать Geany.

**Comment by miks1965** [comment]
На всём зоопарке Linuxов стартует кроме старых версий xUbuntu. Не вижу проблем, почему здесь не стартанёт. Скачивайте отсюда http://pascalabc.net/downloads/PABCNETC.zip и пробуйте )

Для запуска понадобится кодировка 1251 - она там прошита в системном модуле.

**Comment by EntityFX** [comment]
![image](https://user-images.githubusercontent.com/4529778/104316269-47971c80-54ed-11eb-8f0d-1f94d34cee50.png)

Так, стартует.

Осталось понять как его собирать минимально из исходников без тестов и aot.

**Reply by miks1965** [reply]
Ну я не знаю, как там всё на Эльбрусах. 

В папке pabcnetc есть проект PABCNETC.csproj - его надо собрать и по идее это всё.

Или в папке pabcnetcclear - этот консольный компилятор очищен от диалогового режима и раскраски консоли. Именно он используется на Линуксах в олимпиадных системах.

**Reply by EntityFX** [reply]
Да, так сборка работает, единственное нужно понять как Lng и Lib (*.pcu) собираются.

**Reply by EntityFX** [reply]
Пытается искать парсер PascalABCParser.dll, но в master ветке PascalABCParserNewSaushkin

**Reply by SunSerega** [reply]
Проект парсера, вроде, надо 1 раз ручками отдельно собрать, чтоб вся сборка работала.

**Reply by EntityFX** [reply]
> Проект парсера, вроде, надо 1 раз ручками отдельно собрать, чтоб вся сборка работала.

А какой парсер собирать, вижу только https://github.com/pascalabcnet/pascalabcnet/tree/master/Parsers/PascalABCParserNewSaushkin

**Reply by SunSerega** [reply]
Ну, да, а вы ищете что то другое?

**Reply by EntityFX** [reply]
директории Lng и Lib (*.pcu) знаете как формируются?

**Comment by miks1965** [comment]
Так. Я неправильно Вам сказал. В корне проекта есть решение pabcnetc.sln. Его надо собрать. После этого в \bin появятся pabcnetc.exe pabcnetcclear.exe и все необходимые для запуска dll. В частности, PascalABCParser.dll. Всё. Теперь можно подать консольному компилятору pas-файл: pabcnetcclear.exe a.pas. Нужные pcu в Lib автоматически создадутся если их там не было. pcu - это откомпилированные модули. Если модуль не откомпилирован, он компилируется, если откомпилирован, то содержимое pcu разворачивается в память. 

Минимально необходимые для консольного компилятора файлы - на скриншоте: 
![изображение](https://user-images.githubusercontent.com/10325391/104630461-d2257a80-56ab-11eb-9c10-174d6538432b.png)
Всё содержимое папок Lib и Lng надо скопировать

Ещё после сборки компилятора желательно запустить install_pabcnetc.bat  - он запускает ngen для всех сборок, генерируя их нативные образы, что ускоряет запуск. Я не знаю, есть ли такое на Эльбрусах, но если есть, надо запустить.

**Reply by EntityFX** [reply]
1. Я не нашёл, где в git-репозитории находятся Lib, Lng файлы.
2. Нужно будет создать Bash-скрипт с билдилкой или CMake.
3. Пока собрать чистые сборки AnyCpu хватает, так как нужен всего лишь компилятор, остальное -- любой редактор, например, Geany.

**Reply by miks1965** [reply]
1. Ну как не нашли. Вот они: https://github.com/pascalabcnet/pascalabcnet/tree/master/bin
2. В скрипте просто надо скопировать эти файлы в инсталлят
3. В скрипте PascalABCNETConsoleZIP.bat уже есть нужный код. Его можно немного изменить: 
```
cd ..\bin
del ..\Release\PACNETConsole.zip
..\utils\pkzipc\pkzipc.exe -add ..\Release\PABCNETC.zip pabcnetc.exe pabcnetcclear.exe Compiler.dll CompilerTools.dll Errors.dll Localization.dll NETGenerator.dll ParserTools.dll ICSharpCode.NRefactory.dll SemanticTree.dll SyntaxTree.dll TreeConverter.dll OptimizerConversion.dll PascalABCParser.dll licence.txt PascalABCNET.chm System.Threading.dll SyntaxTreeConverters.dll YieldHelpers.dll SyntaxVisitors.dll
..\utils\pkzipc\pkzipc.exe -add -nozip -dir ..\Release\PABCNETC.zip Lib\*.pas
..\utils\pkzipc\pkzipc.exe -add -nozip -dir ..\Release\PABCNETC.zip Lng\*.dat
..\utils\pkzipc\pkzipc.exe -add -nozip -dir ..\Release\PABCNETC.zip Lng\*.LanguageName
..\utils\pkzipc\pkzipc.exe -add -nozip -dir ..\Release\PABCNETC.zip doc\*.*
cd ..\ReleaseGenerators
..\utils\pkzipc\pkzipc.exe -add -nozip -dir ..\Release\PABCNETC.zip install_pabcnetc.bat
```

pcu файлы создадутся при компиляции первой программы с этим модулем.

**Reply by EntityFX** [reply]
Так, ок, понятно.
